File: src/main/java/com/github/kydzombie/link/packet/LinkConnectionsPacket.java
package com.github.kydzombie.link.packet;

import com.github.kydzombie.link.Link;
import com.github.kydzombie.link.gui.LinkTerminalGui;
import com.github.kydzombie.link.util.LinkConnectionInfo;
import net.fabricmc.loader.api.FabricLoader;
import net.minecraft.client.Minecraft;
import net.minecraft.network.PacketHandler;
import net.minecraft.packet.AbstractPacket;
import net.modificationstation.stationapi.api.packet.IdentifiablePacket;
import net.modificationstation.stationapi.api.registry.Identifier;
import org.jetbrains.annotations.NotNull;
import org.lwjgl.util.Color;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;

public class LinkConnectionsPacket extends AbstractPacket implements IdentifiablePacket {
    @NotNull private LinkConnectionInfo[] connections;

    public LinkConnectionsPacket() {
        super();
    }

    public LinkConnectionsPacket(@NotNull LinkConnectionInfo[] connections) {
        super();
        this.connections = connections;
    }

    @Override
    public void read(DataInputStream dataInputStream) {
        try {
            connections = new LinkConnectionInfo[dataInputStream.readInt()];
            for (int i = 0; i < connections.length; i++) {
                connections[i] = new LinkConnectionInfo(
                        Identifier.of(readString(dataInputStream, 100)),
                        readString(dataInputStream, 100),
                        new Color(
                                dataInputStream.readByte(),
                                dataInputStream.readByte(),
                                dataInputStream.readByte()
                        )
                );
            }
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public void write(DataOutputStream dataOutputStream) {
        try {
            dataOutputStream.writeInt(connections.length);
            for (LinkConnectionInfo connection : connections) {
                writeString(connection.type().toString(), dataOutputStream);
                writeString(connection.name(), dataOutputStream);
                dataOutputStream.write(connection.color().getRedByte());
                dataOutputStream.write(connection.color().getGreenByte());
                dataOutputStream.write(connection.color().getBlueByte());
            }
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public void apply(PacketHandler arg) {
        var screen = ((Minecraft)FabricLoader.getInstance().getGameInstance()).currentScreen;
        if (screen instanceof LinkTerminalGui gui) {
            gui.updateConnections(connections);
        }
    }

    @Override
    public int length() {
        int size = 4;
        for (var connection :
                connections) {
            size += connection.type().toString().length();
            size += connection.name().length();
            size += 3;
        }
        return size;
    }

    @Override
    public Identifier getId() {
        return Link.MOD_ID.id("link_connections");
    }
}

File: src/main/java/com/kydzombie/link/packet/LinkConnectionsPacket.kt
package com.kydzombie.link.packet

import com.kydzombie.link.Link
import com.kydzombie.link.gui.LinkTerminalGui
import com.kydzombie.link.util.LinkConnectionInfo
import net.fabricmc.loader.api.FabricLoader
import net.minecraft.client.Minecraft
import net.minecraft.network.PacketHandler
import net.minecraft.packet.AbstractPacket
import net.modificationstation.stationapi.api.network.packet.IdentifiablePacket
import net.modificationstation.stationapi.api.util.Identifier
import java.io.DataInputStream
import java.io.DataOutputStream

class LinkConnectionsPacket() : AbstractPacket(), IdentifiablePacket {

    private lateinit var connections: Array<LinkConnectionInfo>

    constructor(connections: Array<LinkConnectionInfo>) : this() {
        this.connections = connections
    }

    override fun read(dataInputStream: DataInputStream) {
        connections = Array(dataInputStream.readInt()) {
            LinkConnectionInfo.createFrom(dataInputStream)
        }
    }

    override fun write(dataOutputStream: DataOutputStream) {
        dataOutputStream.writeInt(connections.size)
        for (connection in connections) connection.writeTo(dataOutputStream)
    }

    override fun apply(arg: PacketHandler?) {
        val screen = (FabricLoader.getInstance().gameInstance as Minecraft).currentScreen
        (screen as? LinkTerminalGui)?.updateConnections(connections)
    }

    override fun length(): Int {
        var size = 4
        for (connection in connections) size += connection.size()
        return size
    }

    override fun getId(): Identifier {
        return Link.NAMESPACE.id("link_connections")
    }
}
